### PR TITLE
8335991: Implement Simple Source Files and Instance Main Methods (Fourth Preview)

### PR DESCRIPTION
**Pull request by lahodaj**
This is a partial implementation of JEP 495 - adjustments of the JEP metadata in `PreviewFeature`. There are no language changes associated with this JEP. Changes to the `java.io.IO` class are covered by https://github.com/openjdk/jdk/pull/21693.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a JEP request to be targeted
- [x] Change requires CSR request [JDK-8341725](https://bugs.openjdk.org/browse/JDK-8341725) to be approved

### Issues
 * [JDK-8335991](https://bugs.openjdk.org/browse/JDK-8335991): Implement Simple Source Files and Instance Main Methods (Fourth Preview) (**Sub-task** - P4)
 * [JDK-8335984](https://bugs.openjdk.org/browse/JDK-8335984): JEP 495: Simple Source Files and Instance Main Methods (Fourth Preview) (**JEP**)
 * [JDK-8341725](https://bugs.openjdk.org/browse/JDK-8341725): Implement Simple Source Files and Instance Main Methods (Fourth Preview) (**CSR**)

### Reviewers
 * [Adam Sotona](https://openjdk.org/census#asotona) (@asotona - **Reviewer**)
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21787/head:pull/21787` \
`$ git checkout pull/21787`

Update a local copy of the PR: \
`$ git checkout pull/21787` \
`$ git pull https://git.openjdk.org/jdk.git pull/21787/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21787`

View PR using the GUI difftool: \
`$ git pr show -t 21787`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21787.diff">https://git.openjdk.org/jdk/pull/21787.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21787#issuecomment-2447086922)
</details>

### CODE REVIEW

**Review by asotona — APPROVED** [review]
Looks good to me.

File: src/java.base/share/classes/jdk/internal/javac/PreviewFeature.java [hunk 1]
```diff
@@ -68,7 +68,7 @@ public enum Feature {
         // keeping the constant of a feature that has been integrated or dropped, serves the purpose of muting such warnings.
 
         //---
-        @JEP(number=477, title="Implicitly Declared Classes and Instance Main Methods", status="Third Preview")
+        @JEP(number=495, title="Simple Source Files and Instance Main Methods", status="Fourth Preview")
         IMPLICIT_CLASSES,
```
Comment (jaikiran) [inline src/java.base/share/classes/jdk/internal/javac/PreviewFeature.java:72]: Hello Jan, although it's just an internal enum name, do you think we should rename it to be closer to the JEP's title or is it not worth it?

**Review by jaikiran — APPROVED** [review]
This looks OK to me and matches what is stated in the CSR.

File: src/java.base/share/classes/jdk/internal/javac/PreviewFeature.java (hunk 1 above)
Comment (AlanBateman) [reply src/java.base/share/classes/jdk/internal/javac/PreviewFeature.java:72]: The title also shows up in list of preview APIs in the javadoc so changing seems correct.

Comment (lahodaj) [reply src/java.base/share/classes/jdk/internal/javac/PreviewFeature.java:72]: Please note the title is adjusted to match the current name of the JEP. So the built javadoc should not be an issue. What Jaikiran is speaking about is the enum constant name, which hopefully(!) does not have a meaning outside of the JDK.

We could change the constant to `SIMPLE_SOURCE_FILES_AND_INSTANCE_MAIN_METHODS` (to precisely reflect the JEP's name), but it seems like an unnecessary git churn to me. We might also need to keep the existing constant due to bootstrap issues, then remove it when JDK 25 is switched to bootstrap on JDK 24.

I can do that, if really needed, though.

Comment (jaikiran) [reply src/java.base/share/classes/jdk/internal/javac/PreviewFeature.java:72]: > What Jaikiran is speaking about is the enum constant name, which hopefully(!) does not have a meaning outside of the JDK.

I had looked around the rendered javadoc and a few other places and I didn't notice the enum name showing up anywhere publically. It merely appears to be an internal reference within the code.

**Pull request closed without merging**